### ISSUE TITLE
Encrypt MiddleMan service with AES

### ISSUE THREAD

**Issue by pdnelson** [YCPRadioTelescope/YCP-RT-ControlRoom#167]
Right now, we're sending commands to the telescope in plaintext, unencrypted. What this will do is encrypt that text so we can't be hacked.

This will require communication with any teams that communicate with the Control Room through the MiddleMan service (Front End, Mobile). Those teams will have to encrypt their data using the AES library that corresponds to their language (see references), and we will have to decrypt that data on the Control Room side so we can then use the command. Both sides must use the same AES settings; this is something we must coordinate.

A change we will have to make on both the sending and receiving ends of the MiddleMan is that we will have to send the data through a `Byte[]` type as opposed to a `String`.

Implementation (Update 21 September 2020):
- In `RemoteListener.cs`, all plain-text commands would still be visible by the user; however, the `String data` parameter of `processMessage` would become `Byte[] data`, then the decryption from `Byte[]` to `String` would occur on the line directly under the `processMessage` function declaration so that all the `String` commands could be used as they are now.
  - I imagine the mobile team and front end could follow a similar implementation: Have a `String` capture the command that needs to be sent and then encrypt it to `Byte[]` right before it gets sent.
- Each command being sent from front end/mobile will need to be date/time stamped. The reason is so that in the event of an interception occurring, a command used one day could not be used the next. A command being sent will be able to be used for the next 30 seconds, and after that, will have expired.
- The key will be derived from a binary file called `AESKey.bin` that will be located in the same folder as the main AES file (name/location not yet decided).
  - Similarly, the mobile and front end teams will likely need to use this same model, having the key saved in an `AESKey.bin` file somewhere so that it can be easily updated in the event of a key change (#175).

~~A challenge is how to go about the key:~~
- ~~One way of going about this is to have a static key that both sides will use, and it will need changed manually every so often (less secure, easier to implement)~~
- ~~The key changes every time data is sent (more secure, more difficult to implement)~~

***
References:
- AES for C#: https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.aes?view=netcore-3.1 (control room)
- AES for Node/JS: https://www.npmjs.com/package/aes-js (front end)
- AES for React: https://www.npmjs.com/package/react-native-aes-crypto (mobile)

**Comment by TUllery** [comment]
Instead of everyone implementing AES in each component, how about creating one component put into a single library that everyone uses. The only field that would have to be encrypted is the one that identifies this as a valid command. The rest of the fields can be plain text. The same new component would have the code to validated the key at the destination.

**Comment by pdnelson** [comment]
So, the way I was thinking of implementing this (which I admittedly did not go into much detail in the issue) is similar to what you're describing. In the code, you would still see all of the plain-text commands that would be readable. The only thing that would change is the `String data` at the top would become `Byte[] data`, then the conversion from `Byte[]` to `String` would occur on the line directly under the `processMessage` function declaration.

I was thinking the mobile team and front end could implement it similarly: Rather than encrypt every individual command, have a `String` capture the command that needs to be sent and then convert it to `Byte[]` right before it gets sent.

I may be misunderstanding what you said; if so, I'm free all day today if you would like to talk about it over Zoom!

**Comment by TUllery** [comment]
I am okay with your scenario. Is there something in the encoding that is time specific so a command sent (and recorded by the evil person) today cannot be used next week?

**Comment by pdnelson** [comment]
I like the time-specific idea; that will negate the need for a key that changes on every command, so it can be more secure and easier to implement. I will add a timestamp of some sort in the encoding, and then a check in the decoding that makes sure it matches.

**Comment by pdnelson** [comment]
Turns out we were already receiving a `byte[]` array of data! This means instead of doing a straight conversion from `byte[]` to `String` like we were originally doing, this conversion will be replaced by the `Decrypt` function I wrote using AES.

An additional 'verifyMessage` function is being written that will verify the timestamp and then decide whether the command is able to be used or not. The way this will be implemented depends on what the mobile team and we decide, and it is currently undergoing discussion.
